### PR TITLE
NNX: fix Linen-parity gaps + unit tests

### PR DESCRIPTION
**Pull request by ecnal-cienet**
# Description
NNX (`pure_nnx=True`) had Linen-only / silently-divergent gaps across train / loss / decoder / metrics / GRPO. This closes them and adds correctness unit tests. The fixes apply on `main` independently; PR #3526 (flip defaults to NNX) makes them the live default path, and the UTs pin the behavior either way.

## Fixes

| # | Issue |
|---|-------|
| 1 | `skip_step_on_spikes`: silent no-op on NNX (`apply_gradients` didn't forward `loss`/`grad_norm`). Now forwarded + metric surfaced. |
| 2 | `loss_fn`: NNX checked vocab-tiling before indexer warm-up; reordered to match Linen. |
| 3 | NNX decoder logits guards used `self.model_mode` instead of the call-arg `model_mode`. |
| 4 | `routed_bias`: updates silently dropped on NNX (Linen `"intermediates"` prefix absent on NNX dict). Now matched by suffix. |
| 5 | `record_internal_nn_metrics`: KeyError on NNX. Now NNX-aware via suffix collection. |
| 6 | qwix: crashed under `pure_nnx` with the bridged decoder; bridge now skips qwix's non-`Variable` attrs + a config guard rejects bridged-decoder+qwix. |
| 7 | `maxengine.set_engine_vars_from_base_engine`: AttributeError on NNX; now uses `get_kv_cache_annotations_nnx`. |
| 8 | GRPO `gradient_accumulation_steps>1`: NotImplementedError on NNX. Implemented; also fixed the GA loss metric (`sum/GA`, not `sum/total_weights`). |
| 9 | GRPO `scan_layers=False`: NotImplementedError on NNX. Guard removed (NNX policy already matches the inference layout). |
| 10 | GRPO `optimizer_memory_host_offload`: ignored on NNX; now moves opt state to device before the update. |

Also re-declared the legacy GRPO config fields (`inference_replicas` / `inference_devices_per_replica` /
`inference_rollouts` / `use_pathways_reshard`) in `types.py` — they were dropped from the schema so
`grpo.yml` couldn't load (pre-existing on `main`).

## Tests

`tests/unit/{train_nnx_test,grpo_nnx_test,maxengine_nnx_test,nnx_quant_guard_test}.py` — 27 pass on CPU:

```bash
PYTHONPATH=src JAX_PLATFORMS=cpu python3 -m pytest \
  tests/unit/grpo_nnx_test.py tests/unit/train_nnx_test.py \
  tests/unit/maxengine_nnx_test.py tests/unit/nnx_quant_guard_test.py -q
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/experimental/rl/grpo_trainer.py [hunk 1]
```diff
@@ -505,13 +497,61 @@ def diff_wrapper(param, rest, ref_state, config, data):
...
+    }
+    carry, aux = jax.lax.scan(accumulate_gradient, init_carry, ga_data, length=config.gradient_accumulation_steps)
+    # total_loss is already a per-batch mean (and includes moe_lb), so the full-batch
+    # loss is the mean across the equal-sized microbatches.
+    loss = carry["loss"] / config.gradient_accumulation_steps
+    raw_grads = jax.tree_util.tree_map(lambda arr: arr / carry["total_weights"], carry["grad"])
+    aux = jax.tree.map(lambda x: jnp.sum(x, axis=0), aux)
+    new_rest = carry["rest"]
+  else:
+    (loss, (aux, new_rest)), raw_grads = grad_func(curr_params, rest, ref_state, config, data)
```
Comment (igorts-git) [inline src/maxtext/experimental/rl/grpo_trainer.py]: Nit, I think the code would be more readable if it was written such that this short part comes first:
```
if config.gradient_accumulation_steps <= 1:
    (loss, (aux, new_rest)), raw_grads = grad_func(curr_params, rest, ref_state, config, data)
else:
     # The big code block here
```

File: src/maxtext/trainers/pre_train/train.py [hunk 2]
```diff
@@ -208,15 +208,15 @@ def loss_fn(model, config, data, dropout_rng, params, sparsity_state=None, is_tr
       intermediate_outputs["mtp_losses"] = nnx.pop(model, mtp_losses).to_pure_dict()
       intermediate_outputs["mtp_acceptance"] = nnx.pop(model, mtp_acceptance).to_pure_dict()
 
-    if config.num_vocab_tiling > 1:
-      hidden_state_key = ("decoder", "hidden_states")
-      hidden_states = maxtext_utils.get_nested_value(intermediate_outputs, hidden_state_key)[0]
-      xent_sum, total_z_loss = vocab_tiling_nnx_loss(model, hidden_states, data, config, is_train)
-    elif (config.use_indexer and not config.indexer_sparse_training) and is_train:
+    if (config.use_indexer and not config.indexer_sparse_training) and is_train:
```
Comment (NuojCheng) [inline src/maxtext/trainers/pre_train/train.py:211]: Could you help me understand why change the order of `if config.num_vocab_tiling > 1` and `if (config.use_indexer and not config.indexer_sparse_training) and is_train`?

File: src/maxtext/trainers/pre_train/train.py [hunk 3]
```diff
@@ -293,8 +293,25 @@ def loss_fn(model, config, data, dropout_rng, params, sparsity_state=None, is_tr
   # get MoE routed bias term updates
   moe_bias_updates = None
   if config.routed_bias and config.routed_bias_update_rate > 0.0:
-    nested_key = ("intermediates", "decoder", "moe_layers", "moe_bias_updates")
-    moe_bias_updates = maxtext_utils.get_nested_value(intermediate_outputs, nested_key, None)
+    if isinstance(model, nn.Module):
```
Comment (NuojCheng) [inline src/maxtext/trainers/pre_train/train.py:296]: does this branch get triggered only when `enable_nnx=False`?

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/trainers/pre_train/train.py (hunk 3 above)
Comment (ecnal-cienet) [reply src/maxtext/trainers/pre_train/train.py:296]: `isinstance(model, nn.Module)`is the Linen path — the same predicate this function uses for the Linen/NNX split throughout (the forward at the top, etc.); the `else` is the NNX path. It keys on the model type, not a flag directly: a pure-NNX model (`pure_nnx=True`, built with `rngs` → `models.Transformer`) takes the `else`; everything else is a Linen `nn.Module` and takes this branch. So it's `pure_nnx=False`, which is broader than `enable_nnx=False` — it also covers `enable_nnx=True, pure_nnx=False`, where `transformer_as_linen` bridges the NNX core into a Linen `nn.Module`.

File: src/maxtext/trainers/pre_train/train.py (hunk 2 above)
Comment (ecnal-cienet) [reply src/maxtext/trainers/pre_train/train.py:211]: This restores parity with the Linen branch just above (lines 152–160), which checks the indexer dense warm-up before `num_vocab_tiling > 1`; the NNX branch had them reversed. The two conditions are mutually exclusive, so the order only matters when both are enabled (an indexer model trained with vocab tiling): during dense warm-up the main LM loss must be skipped entirely (`xent_sum = 0` — main params frozen, only the indexer trains via KL). With tiling checked first, a warm-up step would run the vocab-tiling loss instead of skipping it. Putting the warm-up check first gives it precedence and matches Linen.

**Review by igorts-git — APPROVED** [review]

File: src/maxtext/experimental/rl/grpo_trainer.py (hunk 1 above)
Comment (ecnal-cienet) [reply src/maxtext/experimental/rl/grpo_trainer.py]: Fixed. Thank you.
